bin/main

```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env ruby
+
+# frozen_string_literal: true
+
+# rubocop:disable Metrics/PerceivedComplexity
+
+# rubocop:disable Metrics/CyclomaticComplexity
+
+# rubocop:disable Metrics/ModuleLength
+
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    entry = is_a?(Range) ? to_a : self
+    i = 0
+    while i < size
+      yield(self[i])
+      i += 1
+    end
+    entry
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    entry = is_a?(Range) ? to_a : self
+
+    i = 0
+    while i < size
+      yield(self[i], i)
+      i += 1
+    end
+    entry
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    i = 0
+    arr = []
+    while i < size
+      arr << self[i] if yield(self[i])
+      i += 1
+    end
+    arr
+  end
+
+  def my_all?(pattern = nil)
+    unless block_given?
+      return my_all? { |obj| obj } unless pattern
+
+      if pattern.class == Regexp
+        my_all? { |x| pattern.match(x) }
+      elsif pattern.class == Class
+        my_all? { |x| x.is_a? pattern }
+      else
+        my_all? { |x| x === pattern }
+      end
+    end
+    my_each { |x| return false unless yield x }
+    true
+  end
+
+  def my_any?(pattern = nil)
+    unless block_given?
+      return my_any? { |obj| obj } unless pattern
+
+      if pattern.class == Regexp
+        my_any? { |x| pattern.match(x) }
+      elsif pattern.class == Class
+        my_any? { |x| x.is_a? pattern }
+      else
+        my_any? { |x| x === pattern }
+      end
+    end
+    my_each { |x| return true if yield x }
+    false
+  end    
+
+  def my_none?(pattern = nil)
+    if block_given?
+      !my_any? { |x| yield x }
+    elsif pattern
+      !my_any?(pattern)
+    else
+      my_each { |x| return false if x }
+      true
+    end
+  end
```

Comment:
- Same issues with my_any? and my_all?.

Example of the errors:
```ruby
irb(main):003:0> arr = [1,5,8,7,4,8]
=> [1, 5, 8, 7, 4, 8]
irb(main):004:0> arr.my_none?
=> false
irb(main):005:0> arr.my_none?(Integer)
Traceback (most recent call last):
        8: from /home/devmod/.rbenv/versions/2.6.0/bin/irb:23:in `<main>'
        7: from /home/devmod/.rbenv/versions/2.6.0/bin/irb:23:in `load'
        6: from /home/devmod/.rbenv/versions/2.6.0/lib/ruby/gems/2.6.0/gems/irb-1.0.0/exe/irb:11:in `<top (required)>'
        5: from (irb):5
        4: from bin/main.rb:85:in `my_none?'
        3: from bin/main.rb:77:in `my_any?'
        2: from bin/main.rb:18:in `my_each'
        1: from bin/main.rb:77:in `block in my_any?'
LocalJumpError (no block given (yield))
```

Reply:
I honestly don't know any other way to solve for these scenarios.